src/index.ts

```diff
@@ -0,0 +1,316 @@
+/**
+ * @fileOverview Exports the Sajari javascript api.
+ * @name index.ts
+ * @author Sajari
+ * @license MIT
+ * @module sajari
+ */
+
+import {
+  newRequest,
+  newAggregates,
+  newQueryID,
+  newResult,
+  newResults,
+  valueFromProto
+} from "./constructors";
+import { newError, SearchError } from "./error";
+import { AggregateResponse, Result, Results, ResultValues } from "./results";
+
+const UserAgent = "sdk-js-1.0.0";
+
+export interface Tracking {
+  // Tracking specifies which kind (if any) tokens should be generated and returned
+  // with the query results.
+  type: TrackingType;
+
+  // QueryID is a unique identifier for a single search query.  In the
+  // case of live querying this is defined to be multiple individual queries
+  // (i.e. as a user types the query is re-run).
+  query_id: string;
+
+  // Sequence (i.e. sequential identifier) of this  in the context of a
+  // sequence of queries.
+  sequence: number;
+
+  // Field is the field to be used for adding identifier information to
+  // generated tokens (see TrackingType).
+  field: string;
+
+  // Data are values which will be recorded along with tracking data produced
+  // for the request.
+  data: Values;
+}
+
+/**
+ * Values is a plan object with only string values.
+ */
+export interface Values {
+  [id: string]: string;
+}
+
+export interface Session {
+  next(values: Values): [Tracking, undefined] | [undefined, Error];
+  reset(): void;
+}
+
+export const TrackingNone: string = "NONE";
+export const TrackingClick: string = "CLICK";
+export const TrackingPosNeg: string = "POS_NEG";
+
+export const enum TrackingType {
+  TrackingNone = "NONE",
+  TrackingClick = "CLICK",
+  TrackingPosNeg = "POS_NEG"
+}
+
+/**
+ * TextSession creates a session based on text searches.
+ * It automatically resets once the value specified by the query label has changed in certain ways.
+ */
+export class TextSession implements Session {
+  private queryLabel: string;
+  private session: Session;
+  private lastQuery: string = "";
+
+  public constructor(queryLabel: string, session: Session) {
+    this.queryLabel = queryLabel;
+    this.session = session;
+  }
+
+  public next(values: Values): [Tracking, undefined] | [undefined, Error] {
+    const text = values[this.queryLabel];
+    if (text === undefined) {
+      this.reset();
+      return this.session.next(values);
+    }
+
+    const shortenedPrevQ = this.lastQuery.substr(0, Math.min(text.length, 3));
+    const first3CharactersChanged = !(
+      text.substr(0, shortenedPrevQ.length) === shortenedPrevQ
+    );
+    const queryCleared = this.lastQuery.length > 0 && text.length === 0;
+    if (first3CharactersChanged || queryCleared) {
+      this.reset();
+    }
+    this.lastQuery = text;
+
+    return this.session.next(values);
+  }
+
+  public reset(): void {
+    this.session.reset();
+  }
+}
+
+/**
+ * Session holds state about a series of searches.
```

Comment:
BaseSession holds...

Also: series, sequence... let's be consistent and make clear when they mean the same thing, or don't mean the same thing.